### PR TITLE
Fix and clarify optional license emitting

### PR DESCRIPTION
**Pull request by kripken**
Alternative to #11259, adds testing and a fix to the default
license header from there, which was wrong in `shared.py`
compared to what is in `js/`.

This clarifies that `EMIT_EMSCRIPTEN_LICENSE` does
nothing in `-O0` and `-O1` - those modes do not do any JS
opts, and leave all comments in the JS. License comments
are comments like all others, and may appear, and may
appear more than once.

In `-O2+` ensure that we emit a single license header, and
only if requested. We had two bugs here:

* Closure would combine licenses with different years.

* Only noticeable internally, but when running some passes
with `noPrint` (like metadce info gathering) we do not print
any JS, and so should not print a license.

### CODE REVIEW

File: tests/test_other.py [hunk 1]
```diff
@@ -9230,23 +9230,34 @@ def test_emscripten_metadata(self):
...
     run_process([PYTHON, EMCC, path_from_root('tests', 'hello_world.c')] + args)
     with open('a.out.js') as f:
       js = f.read()
-    self.assertContainedIf('Copyright 2010 Emscripten authors', js, expect_license)
-    self.assertContainedIf('SPDX-License-Identifier: MIT', js, expect_license)
-    if expect_license:
-      self.assertEqual(js.count('Copyright 2010 Emscripten authors'), 1)
+    licenses_found = len(re.findall('Copyright [0-9]* The Emscripten Authors', js))
+    if expect_license and licenses_found == 0:
+      self.fail('Unable to find license block in output file!')
```
Comment (sbc100) [inline tests/test_other.py:9256]: Probably better written as:

```
if expect_license:
    self.assertEqual(licenses_found, 0, 'Unable to find license block in output file!')
```

Any time you see `self.fail` in normal test code its probably a sign something could be improved (IMHO).

**Comment by sbc100** [comment]
Seems a little complicated but I don't have any immediate thoughts on how to reduce that.

**Comment by juj** [comment]
> This clarifies that `EMIT_EMSCRIPTEN_LICENSE` does
> nothing in `-O0` and `-O1` - those modes do not do any JS
> opts, and leave all comments in the JS. License comments
> are comments like all others, and may appear, and may
> appear more than once.

I dislike this since it will leave the license headers in the beginning of the file in -O0 and -O1 builds. This still in my opinion gives the wrong impression as if anything that is outputted by Emscripten would have to be MIT-licensed, with copyright to the output owned by Emscripten Authors.

By extension we have a problem at Unity that people doing debug/iteration builds of their Unity games to the web would have a message in the beginning of their game file saying copyright The Emscripten Authors, license MIT. This might make the game developers misunderstand, or the game players misunderstand that either their game copyright would be owned by someone else than them as the game creator, or that rights to use the Unity web export support might be owned by someone else than their Unity license.

I believe we will either have to limit to not supporting -O0/-O1 builds, or given that there is no attribution requirement, further postprocess the files to drop the confusing license headers from the beginning of the build output.

**Comment by sbc100** [comment]
I assume that stripping all comments in O0 and O1 would be undesirable?

Can we instead have `EMIT_EMSCRIPTEN_LICENSE=0` explicitly look for and strip out any license headers?

**Comment by sbc100** [comment]
We should also take some action to confirm that the generated JS is indeed not converted by out MIT.

**Comment by kripken** [comment]
I see @juj. Ok, I rewrote this to address that. This replaces all the logic with a single place at the end, and should work for all build types.

The downside is that this does an extra write of the JS. But it's probably not noticeable, and I don't see any other way to address all the requirements from @juj and @hujiajie both here and in #11255

**Comment by juj** [comment]
lgtm

**Pull request merged**